Added the safe-to-ignore tag to version_edit

zhichao-cao opened the pull request:
Each time RocksDB switches to a new MANIFEST file from old one, it calls WriteCurrentStateToManifest() which writes a 'snapshot' of the current in-memory state of versions to the beginning of the new manifest as a bunch of version edits. We can distinguish these version edits from other version edits written during normal operations with a custom, safe-to-ignore tag.

Test plan: added test to version_edit_test, pass make asan_check

riversand963 commented on db/version_edit.cc:
```diff
@@ -59,6 +59,7 @@ enum Tag : uint32_t {
   kDbId,
   kBlobFileAddition,
   kBlobFileGarbage,
+  kSafeToIgnore,
```
Sorry for the confusion during offline communication. I suggest that we rename it to something like `kStateUponManifestSwitch`.

zhichao-cao commented:
> Thanks @zhichao-cao for working on this. I wonder whether an end marker is needed to denote the end of such version edits written when switching MANIFESTs.

Thanks for the review. The new commit addresses the comments.

riversand963 commented on db/version_edit.h:
```diff
@@ -423,6 +423,19 @@ class VersionEdit {
+  bool GetManifestSwitchFinishedTag() const {
+    return manifest_switch_finished_;
+  }
+
```
For this group of functions, I suggest that we rename them by removing `SwitchTag` or `FinishedTag`.

zhichao-cao replied:
@riversand963 I will rename to SetStateUponManifest/GetStateUponManifest and SetManifestSwitched/GetManifestSwitched

riversand963 commented on db/version_edit.h:
```diff
@@ -470,6 +483,13 @@ class VersionEdit {
   bool is_in_atomic_group_ = false;
   uint32_t remaining_entries_ = 0;
+  // To distinguish the version edit written by WriteCurrentStateToManifest
+  // and other redgular writes. Default is false.
```
Nit: typo regular

riversand963 commented on db/version_set.cc:
```diff
@@ -5072,6 +5075,18 @@ Status VersionSet::WriteCurrentStateToManifest(
+  Status s_end_record = log->AddRecord(end_record);
+  if (!s_end_record.ok()) {
+    return s_end_record;
+  }
```
How about just returning `s_end_of_record` directly.

zhichao-cao replied:
@riversand963 Sure
